hypothesis-python/src/hypothesis/internal/conjecture/datatree.py

```diff
@@ -204,22 +286,83 @@ def generate_novel_prefix(self, random):
+                value = float_to_int(value)
+            return (value, cd.buffer)
+
+        def draw_buf(ir_type, kwargs, *, forced):
+            (_, buf) = draw(ir_type, kwargs, forced=forced)
+            return buf
+
+        def draw_value(ir_type, kwargs):
+            (value, _) = draw(ir_type, kwargs)
+            return value
+
+        def append_value(ir_type, kwargs, *, forced):
+            novel_prefix.extend(draw_buf(ir_type, kwargs, forced=forced))
+
+        def append_buf(buf):
+            novel_prefix.extend(buf)
 
         current_node = self.root
         while True:
             assert not current_node.is_exhausted
-            for i, (n_bits, value) in enumerate(
-                zip(current_node.bit_lengths, current_node.values)
+            for i, (ir_type, kwargs, value) in enumerate(
+                zip(current_node.ir_types, current_node.kwargs, current_node.values)
             ):
                 if i in current_node.forced:
-                    append_int(n_bits, value)
+                    if ir_type == "float":
+                        value = int_to_float(value)
+                    append_value(ir_type, kwargs, forced=value)
                 else:
+                    attempts = 0
                     while True:
-                        k = random.getrandbits(n_bits)
-                        if k != value:
-                            append_int(n_bits, k)
+                        (v, buf) = draw(ir_type, kwargs)
+                        if v != value:
+                            append_buf(buf)
+                            break
+
+                        # it may be that drawing a previously unseen value here is
+                        # extremely unlikely given the ir_type and kwargs. E.g.
+                        # consider draw_boolean(p=0.0001), where the False branch
+                        # has already been explored. Generating True here with
+                        # rejection sampling could take many thousands of loops.
+                        #
+                        # If we draw the same previously-seen value more than 5
+                        # times, we'll go back to the unweighted variant of the
+                        # kwargs, depending on the ir_type. Rejection sampling
+                        # produces an unseen value here within a reasonable time
+                        # for all current ir types - two or three draws, at worst.
+                        attempts += 1
+                        if attempts > 5:
+                            kwargs = {
+                                k: v
+                                for k, v in kwargs.items()
+                                # draw_boolean: p
+                                # draw_integer: weights
+                                if k not in {"p", "weights"}
+                            }
+                            while True:
+                                (v, buf) = draw(ir_type, kwargs)
```

Comment:
I wrote this because I had initially convinced myself that we should try to follow the input distribution specified by the programmer when generating novel prefixes, and this is a middle ground between "follow the distribution exactly" (using `kwargs` unmodified) and "use the neutral distribution from the start" (deleting `p` and `weights` from `kwargs` from the start, don't even track attempts). 

But I've been thinking about it more, and I'm not sure we want to follow the programmer's input distribution in novel prefixes. We do on current master: without an ir to help with deduplication, there may be 1 way to represent True and 10k to represent False for a single draw_boolean, and random bit drawing is going to put you in the latter bucket almost every time. But we didn't have any mechanism to *not* follow the programmer's distribution on master, it just fell out of dealing with bits.

My concern was that if the programmer had specified a particular distribution for generating inputs, and we instead generate novel prefixes uniformly, we skew away from their distribution and towards uniform. But if we're only doing this for novel prefixes then maybe it doesn't really matter? We generate something novel and let the random bit draws afterwards conform to the programmer's distribution...

Reply:
I guess I'm also worried about not following our own internal distributions.

Let me take a step back: say you have a strategy which draws value $NORMAL with high probability, and some value $RARE with low probability. Would you expect hypothesis to explore both $NORMAL and $RARE equally, or to bias towards $NORMAL? 

This is a concern anywhere `draw_boolean(p=...)` or `draw_integer(weights=...)` is used internally. A contrived example:

```python
from hypothesis import *
from hypothesis.strategies import *

class almost_always_zero(SearchStrategy):
    def do_draw(self, data):
        return 0 if data.draw_boolean(0.99) else 1

@given(data())
def f(data):
    n = data.draw(almost_always_zero())
    data.draw(integers()) # avoid exhausting tree
    print(n)
f()
```

This exploration happens in two places inside `generate_novel_prefix`; drawing an initial value for an as-of-yet unexplored node, and picking which child to explore during a transition. We probably want consistent behavior across these two locations, but this isn't necessarily a requirement.

Drawing using `kwargs` corresponds to biasing towards $NORMAL (current master behavior), while removing `p` and `weights` from `kwargs` before drawing corresponds to exploring $NORMAL and $RARE equally.

(and then there is the added complication of, if we do draw using unmodified kwargs, the probability may be almost impossible to satisfy, so we should give up and stray from the distribution at some point for performance; see above code for an attempt at that).

Reply:
Writing this comment out helped clarify my thinking, and I'm sure I've contradicted a bunch of what I said in my original comment in this chain. I'm pretty convinced that we want to bias towards $COMMON here in the distribution in all cases — otherwise what's the point in having a `p` argument?

(and I'm not sure that you ever disagreed with this conclusion either, given how muddy my original comment was.)

This means we'll also need some anti-performance-ruining code. The `if attempts > 5:` code is doing the right thing in spirit here, but we can actually do a lot better than rejection sampling by redistributing the probabilities of the already chosen branch. That would require some additional code per ir node type. I think it's worthwhile.

Reply:
The example that clarified my thinking, and inclined me towards respecting the distribution:  suppose that we have a tuple consisting of `draw_boolean(p=0.99)` and `draw_integer()`.  Do we really want the distribution to be radically different based on the order of two independent draws?

On the other hand, I think people make low probabilities _way too low_ for efficient bug-finding, so as usual we should probably try a mixed strategy of some kind (e.g.: for each test case, choose a meta-probability with which to use the specified probabilities).  See e.g. [this paper](https://agroce.github.io/onwardessays21.pdf) or search "divers" in [my literature review here](https://hypofuzz.com/docs/literature.html) for more thoughts.

Reply:
This line of diversity reasoning makes sense, and I think the meta-probability approach wouldn't go amiss. I've left a TODO and will come back to that approach in a future PR.